Implement CRI pause image support

adisky opened the pull request:
This commit adds "Pinned" as "true" in CRI image struct if the image has been configured as sandbox_image in containerd config, also labels the sandbox image in containerd store to avoid additional field in container image store

fixes: https://github.com/containerd/containerd/issues/6352

Signed-off-by: Aditi Sharma <adi.sky17@gmail.com>

adisky commented:
cc @mikebrow @endocrimes

adisky commented:
working on adding integration test

adisky commented:
This is ready for review, added integration tests.

marquiz commented on pkg/cri/util/image.go:
```diff
@@ -31,3 +39,8 @@ import (
 	return docker.ParseDockerRef(ref)
 }
+
+// GetSandboxImageLabels returns label applied on sandbox image
+func GetSandboxImageLabels() (string, string) {
+	return sandboxImageLabelKey, sandboxImageLabelValue
+}
```
Is there any value in specifying this function or would it be just easier and more readable to export (and use) the consts directly(?) 🤔 And, should they be specified in `labels/labels.go` instead?

A helper function like this would make more sense to me:
```suggestion
func IsSandboxImage(labels map[string]string) bool {
	return labels[sandboxImageLabelKey] == sandboxImageLabelValue
}
```

mikebrow reviewed (commented):
I'd like to discuss making this more generic... would like to consider using an image annotation .. For example, when pulling the pause image add an annotation that requests the image be flagged as pinned.

mikebrow commented:
@mrunalp What do you think? 

This kep/feature seems to be allowing a runtime to manage pinning status to suggest to the kubelet to not collect this (a particular) image based on the new pinned flag.  

Eventually I think we should have a more robust container image policy spec, that can be expressed over the image service and/or the runtime service for images used by a pod/container.  Sans having that in hand, and given that this pinned feature is supported in kubelet.  Perhaps we could at least start experimenting by allowing pull requests to indicate a desire for the image to be pinned. Sans adding a new field or policy struct to the pull image request struct maybe just use an image annotation? I think that would be better than people figuring out how to open up the image db in our store and pinning images that way :-)

adisky commented on pkg/cri/server/image_pull.go:
```diff
@@ -113,12 +114,14 @@ func (c *criService) PullImage(ctx context.Context, r *runtime.PullImageRequest)
 		containerd.WithPullSnapshotter(c.config.ContainerdConfig.Snapshotter),
 		containerd.WithPullUnpack,
-		containerd.WithPullLabel(imageLabelKey, imageLabelValue),
+		containerd.WithPullLabels(labels),
```
> I'd like to discuss making this more generic... would like to consider using an image annotation .. For example, when pulling the pause image add an annotation that requests the image be flagged as pinned.

@mikebrow already doing what you suggested, adding an annotation when pause image is pulled, anything i am missing here?
```
func (c *criService) getLabels(name string) map[string]string {
	labels := map[string]string{imageLabelKey: imageLabelValue}
	// Adding sandbox image label to containerd store to avoid
	// new CRI field "Pinned" in containerd image store
	if c.config.SandboxImage == name {
		sandboxImageLabelKey, sandboxImageLabelValue := util.GetSandboxImageLabels()
		labels[sandboxImageLabelKey] = sandboxImageLabelValue
	}
	return labels
}
```

scuzhanglei replied:
```golang
if c.config.SandboxImage == name {
		sandboxImageLabelKey, sandboxImageLabelValue := util.GetSandboxImageLabels()
		labels[sandboxImageLabelKey] = sandboxImageLabelValue
	}
````

current implementation the label name and value is strongly related to  sandbox image,  how about make it more generic, just a label to indicate the image is pinned.

mikebrow replied:
missed your response when I was on vac..  

right.. so what I'm talking about is in PullImage() (inbound CRI pullrequest) check PullImageRequest.Image.Annotations[] for an annotation (maybe "k8s.io/cri/pinned"="true" or similar) indicating that this image when successfully pulled should be pinned

mikebrow replied:
in this fashion.. the caller to PullImage can identify if the image should carry the pinned attribute

on the other side in our store this annotation will already be there.. in the store image.. though you could also add a label for pinned as @scuzhanglei mentioned..

mikebrow replied:
A quick check in kubelet image_manager... the image annotations on the pull image request are the set of pod annotations.. I don't see another meta object injection method in kubelet that apples to images ..

ruiwen-zhao replied:
+1 on making this logic more generic than just sandbox image. I think there are two cases where containerd wants to pin an image:

a) the sandbox image. Containerd will just pin the image if the image name matches the sandbox image name.

b) some arbitrary image that the user wants to pin. The user (caller of pullImage) can add a `PullImageRequest.Image.Annotation` (maybe "k8s.io/cri/pinned"="true" or similar). Containerd will check this Annotation and mark the image as pinned. 

Seems like this PR supports case a only.

adisky commented:
> This kep/feature seems to be allowing a runtime to manage pinning status to suggest to the kubelet to not collect this (a particular) image based on the new pinned flag.
> 
> Eventually I think we should have a more robust container image policy spec, that can be expressed over the image service and/or the runtime service for images used by a pod/container. 

agrees on that

> Sans adding a new field or policy struct to the pull image request struct maybe just use an image annotation? I think that would be better than people figuring out how to open up the image db in our store and pinning images that way :-)
The new field already being added 

little confused here :)  field we have already added in cri-api. Kubelet is expecting container runtime to tell about pinned images as it is specified in the containerd config. Also added the additional in containerd till cri code level, for containerd store used an annotation that is later converted into the field.

mikebrow commented:
> little confused here :) field we have already added in cri-api. Kubelet is expecting container runtime to tell about pinned images as it is specified in the containerd config. Also added the additional in containerd till cri code level, for containerd store used an annotation that is later converted into the field.

right.. and there is no way to "set" an arbitrary image to pinned..

current
```
type PullImageRequest struct {
	// Spec of the image.
	Image *ImageSpec `protobuf:"bytes,1,opt,name=image,proto3" json:"image,omitempty"`
	// Authentication configuration for pulling the image.
	Auth *AuthConfig `protobuf:"bytes,2,opt,name=auth,proto3" json:"auth,omitempty"`
	// Config of the PodSandbox, which is used to pull image in PodSandbox context.
	SandboxConfig        *PodSandboxConfig `protobuf:"bytes,3,opt,name=sandbox_config,json=sandboxConfig,proto3" json:"sandbox_config,omitempty"`
	XXX_NoUnkeyedLiteral struct{}          `json:"-"`
	XXX_sizecache        int32             `json:"-"`
}
```

we could add image cache policies to this pull request struct

eg. pin image until rule (until the image has been used n times, until duration has passed, ...)

Alternatively one could just add a pin field to the pull request, and later on we could add another field for more interesting cache policies

Additionally on image import (currently used primarily for disconnected never pull image policy pods) we could also specify a cache policy there...

adisky commented:
@mikebrow Thanks for your reply, was little busy from past few weeks will go through them and get back to you

ruiwen-zhao commented:
+1 on making the support more generic, i.e. to support user pinning arbitrary images. 

To pin sandbox image, it makes sense for containerd to check whether the image name matches the sandbox image name, because container runtime has the sole responsibility and context about sandbox image.

To support pinning arbitrary images, however, containerd would need to be able to receive a "pinning request". I personally like Mike's suggestion of using `PullImageRequest.Image.Annotation`, to avoid extra fields in PullImageRequest. 
(Do note that with current kubelet's image manager implementaion, pod annotations are copied over to `PullImageRequest.Image.Annotation`, which means that k8s users can pin an image with pod annotations).


It is ok to leave the arbitrary image pinning support to a separate issue/PR. I can go create an issue if needed,

Any thoughts? @mikebrow @mrunalp

seanyen commented:
another aspect to comment here: how does the [lease](https://github.com/containerd/containerd/blob/main/docs/garbage-collection.md) interact with the `pinned` here? should all the images under the lease lock also report as `pinned` state?

One example is that K3s is using the [lease to lock](https://github.com/k3s-io/k3s/blob/v1.23.8%2Bk3s2/pkg/agent/containerd/containerd.go#L166-L190) the local preloaded images for airgap scenario. It seems to only prevent from containerd to recycle those and because kubelet is not sharing the same information, the garbage collection on kubelet could still make a decision to remove them.

SergeyKanzhelev commented:
Reviving this PR. Can we pick a generic name for the label and keep this PR down to implementing the sandbox image pinning? Pinning other images may be done later, I think it is not very controversial decision to use labels to store the pinning status.

adisky commented:
closing this PR in favour of https://github.com/containerd/containerd/pull/7944

adisky commented:
/close
